Make Data.ByteString.Lazy.Char8.lines less strict

vdukhovni opened the pull request:
The current implementation of `lines` in Data.ByteString.Lazy.Char8 is too strict.  When a "line" spans multiple chunks it traverses all the chunks to the first line boundary before constructing the list head.

For example, `lines <$> getContents` reading a large file with no line breaks does not make the first chunk of the (only) line available until the entire file is read into memory.

Now that `Data.ByteString.break` is optimised for the `(== c)` case, we can get efficient code for the common many lines per-chunk use-case, without being needlessly strict.  Tests added to make sure that the first chunk is available prompty without looking further.

hs-viktor commented:
@Bodigrim Something went wrong early in downing GHC 8.2 for Ubuntu, and also cancelled all the other CI jobs. Perhaps a transient glitch, but I have no permissions to kick the CI it seems.  Please make it go. A review would also be great.

clyring reviewed (changes requested):
I retried CI without success. I don't have time to investigate that further right now.

Although it was (partially) documented, the old strictness behavior is surprising and I expect any potential performance benefit is negligible outside of the degenerate small-chunks case, so I'm happy to replace it. (Have you benchmarked to verify that any performance impact is negligible?)

clyring commented on Data/ByteString/Lazy/Char8.hs:
```diff
@@ -856,59 +858,29 @@ unzip :: [(Char, Char)] -> (ByteString, ByteString)
+  where
+    lines1 :: S.ByteString -> ByteString -> NonEmpty ByteString
+    lines1 c cs
+        | len > 1   = Chunk h0 Empty <| lines1 (B.unsafeDrop 1 t0) cs
```
Also, `h0` may be empty, in which case `Chunk h0 Empty` breaks the "no-empty-chunks" invariant.

vdukhovni replied:
This change is not needed or wanted. We DO NOT want to be strict in the tail of the list. The (<|) combinator correctly avoids strictness in the head of the result.

vdukhovni replied:
Good catch on the invariant. Code updated and test added.

clyring replied:
This change does not make us strict in the tail of the list; `(:|)` remains lazy in both arguments.

vdukhovni commented on Data/ByteString/Lazy/Char8.hs:
```diff
@@ -856,59 +858,29 @@ unzip :: [(Char, Char)] -> (ByteString, ByteString)
-            Just n ->
-                let !c' = revChunks (B.unsafeTake n c : line)
-                 in c' : loop0 (B.unsafeDrop (n+1) c) cs
+lines (Chunk c0 cs0) = NE.toList $ lines1 c0 cs0
```
That note is about the strict Data.ByteString, where laziness is not expected.  Here we actually want much laziness, but the combinator used is easily avoided. Pushing an update.

vdukhovni replied:
We need the lazy pattern match for **correctness**.

vdukhovni replied:
OK, I reworked the PR to make sure that lazy constructor fields are internally more strict when the element is forced. Also switched to use `elemIndex` making the control flow more similar to the original.  Tests still pass.  CI should confirm (except for GHC 8.0 and 8.2 which fail to download from the ppr at present and so disabled as a workaround).

vdukhovni replied:
Marking this as resolved. The code is substantially reworked. We should start from scratch. It is short enough...

clyring replied:
(For what it's worth, I was referring to the `let l :| ls = lines1 c0 cs0 in l : ls` that this specific line was updated to, and not referring to the critical explicitly-lazy pattern match below.)

vdukhovni commented on Data/ByteString/Lazy/Char8.hs:
```diff
@@ -856,59 +858,29 @@ unzip :: [(Char, Char)] -> (ByteString, ByteString)
+        | len == 1  = Chunk h0 Empty :| lines cs
+        | otherwise = let ~(l:|ls)    = lines2 cs in Chunk c l :| ls
+      where
+        (h0, t0) = B.break (== 0x0a) c
```
Well, I am not a fan of being pedantic here. The `c2w` function will always return `0x0a` here. But if you insist on matching existing style, I will not get in the way.

clyring replied:
I'm aware of that; it will all get constant-folded to the same thing eventually. And I like not having to think about or remember whether `0x0a` is the right character code for `'\n'`.

vdukhovni replied:
OK, once CI is done for 8.4+, I'll push an update that replaces `0x0a` with `c2w '\n'`. To me they read the same, but I perhaps I'm more accustomed to hex dumps...

vdukhovni replied:
Done.

vdukhovni commented:
We don't have existing benchmarks for lines, but looking at the code path taken when the initial chunk holds multiple lines I see no opportunity for performance degradation.  If anything the new code should be faster, because no `revchunks`, ...

The *in-chunk* loop boils down to:
```haskell
lines (Chunk c0 cs0) = let l :| ls = lines1 c0 cs0 in l : ls
  where 
    lines1 c cs
        | len > 1   = c1 <| lines1 (B.unsafeDrop 1 t0) cs
       ...
      where
        (h0, t0) = B.break (== 0x0a) c
        len      = B.length t0
        c1       = if B.null h0 then Empty else Chunk h0 Empty
```
Which is essentially the same:
```
    loop0 c cs =
        case B.elemIndex (c2w '\n') c of
            ...
            Just n | n /= 0    -> Chunk (B.unsafeTake n c) Empty
                                : loop0 (B.unsafeDrop (n+1) c) cs
                   | otherwise -> Empty
                                : loop0 (B.unsafeTail c) cs
```
Strict ByteStrings optimise `B.break` via RULEs to `B.elemIndex ...` and the rest is the same. The case we optimise for (lines shorter than chunks) wants both h0 and t0 to ultimately be produced, and the new code is simpler and cleaner.

Bodigrim commented:
@vdukhovni please rebase atop of https://github.com/haskell/bytestring/pull/563

vdukhovni commented:
> @vdukhovni please rebase atop of #563

Done.

vdukhovni commented:
Rebased on `master`, now that the CI fix is merged.

clyring commented on Data/ByteString/Lazy/Char8.hs:
```diff
@@ -856,59 +857,52 @@ unzip :: [(Char, Char)] -> (ByteString, ByteString)
+  where
+    -- Force the non-empty list *and* its first element!
+    unNE :: NonEmpty a -> [a]
+    unNE (!a :| b) = a : b
```
I see what you're trying to do here. But we don't just want the head of the list to be evaluated before it sees the light of day: We want it to not be born a thunk in the first place. (...which this patch has enough strictness to achieve! Good!)

And since `a` is definitely not a thunk, the bang here becomes superfluous. And ghc-9.2 and older won't be able to tag-inference the resulting eval into a no-op... so I would recommend removing the bang here.

vdukhovni replied:
Dropped the redundant bang. It did occur to me that it might be redundant.

vdukhovni commented on Data/ByteString/Lazy/Char8.hs:
```diff
@@ -856,59 +857,52 @@ unzip :: [(Char, Char)] -> (ByteString, ByteString)
+                   -- entire remaining chunk was just "\n"
+                   else Empty :| lines cs
+          where
+            c' = Chunk (B.unsafeTake n c) Empty
```
Excellent suggestion, done in a fixup commit.  I'll squash once the reviews are all in.

vdukhovni commented on Data/ByteString/Lazy/Char8.hs:
```diff
@@ -856,59 +857,52 @@ unzip :: [(Char, Char)] -> (ByteString, ByteString)
+        -- first line of result, with the rest of the result lazy!
+        -- In particular, we don't strictly pattern match on 'cs'.
+        Nothing -> let ~(l:|ls) = lazyRest cs
+                    in  Chunk c l :| ls
```
Added comments to that effect.

vdukhovni commented:
@clyring I think is looking pretty clean now. Once there are no residual issues, and you mark the PR approved, I'll squash and wait for any final reviews.

clyring reviewed (commented):
This looks good now. Just two aesthetic quibbles:

* You may wish to remove `Data.ByteString.break` from the commit message.
* The comment for `consNE` is now out-of-date.

The maintainers will generally squash when merging, so doing so yourself is not required.

I did also confirm with a toy benchmark (`nf L8.lines (L.fromStrict $ stimes 32 loremIpsum)`)  that this patch does not regress performance in the typical many-more-lines-than-chunks situation. (...and 827df3043724cbe5e4c1ab57c49e80d1515b1ddd was about 60% slower, almost entirely because `(<|)` is terrible.)

vdukhovni commented:
> This looks good now. Just two aesthetic quibbles:
> 
> * You may wish to remove `Data.ByteString.break` from the commit message.
> * The comment for `consNE` is now out-of-date.

Thanks for pitching in.  I've updated the comments as suggested and squashed. I hope this is it.

sjakobi commented on tests/Properties.hs:
```diff
@@ -82,9 +82,29 @@ prop_unsafeTail xs = not (P.null xs) ==> P.tail xs === P.unsafeTail xs
 prop_unsafeLast xs = not (P.null xs) ==> P.last xs === P.unsafeLast xs
 prop_unsafeInit xs = not (P.null xs) ==> P.init xs === P.unsafeInit xs
 
+prop_lines_empty_invariant =
+     True === case LC.lines (LC.pack "\nfoo\n") of
+        Empty : _ -> True
+        _         -> False
```
Could we also have a randomized test that uses `invariant` or `checkInvariant` to check the output of `lines`? I was expecting this test to exist already, but I couldn't find it.

vdukhovni replied:
Could that be a separate PR, or would you care to propose the test?  It'd take me some scarce cycles to *page in* (or acquire) the knowhow of how to construct randomised tests for this, I don't use the facilities in question sufficiently often in real life...

sjakobi replied:
Yeah, this can be done separately. To generate the test input, I'd simply use some `arbitrary` `LazyByteString`'s, interleaved with additional newlines.

sjakobi replied:
I have opened #564 to track this.

sjakobi reviewed (approved):
Cheers!

vdukhovni commented:
Are we waiting for @Bodigrim to review, or can this now be merged?  By the way, when squashing I did inadvertently merge that `n1 <- n + 1` patch I was asking about. I hope that's OK.

Bodigrim commented:
Thanks @vdukhovni!
